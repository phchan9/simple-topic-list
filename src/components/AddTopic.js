import React, { Component } from 'react';
import PropTypes from 'prop-types';

class AddTopic extends Component {

  state = {
    text: '',
    error: false
  };

  onSubmitHandler = (e) => {
    const { addTopic  } = this.props;
    const text = this.state.text.trim();
    e.preventDefault();

    // blank topic won't be submit
    if (!text.length) {
      return;
    }
    // check if string exceed 255 characters
    else if (text.length > 255) {
      this.setState({ error: true });
    } else {
      addTopic(text);
      this.setState({ error: false });
      this.setState({ text: ''});
    }
  }

  onChangeHandler = (e) => {
    this.setState({ text: e.target.value});
  }

  render() {
    let formInputClass = "form-group " + (this.state.error? "has-error": "");
    let errorMsgLabelClass = "error " + (this.state.error? "": "invisible");

    return (
      <div>
        <form className="form-horizontal" onSubmit={this.onSubmitHandler}>
          <div className={formInputClass}>
            <label>Topic:</label>
            <textarea className="form-control"
                      rows="5"
                      placeholder="A topic doesn't exceed 255 characters."
                      type="text"
                      autoFocus="true"
                      value={this.state.text}
                      onChange={this.onChangeHandler}
              />
          </div>
          <div className="form-group">
            <button type="submit" className="btn btn-default">
              Submit
            </button>
            <label
              className={errorMsgLabelClass}> * A topic doesn't exceed 255 characters
            </label>
          </div>
        </form>
      </div>
    );
  }
}

AddTopic.propTypes = {
  addTopic: PropTypes.func.isRequired
};

export default AddTopic;
