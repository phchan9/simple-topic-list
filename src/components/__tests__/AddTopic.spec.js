import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import AddTopic from '../AddTopic';

const setup = () => {
  const props = {
    addTopic: jest.fn()
  };

  const wrapper = mount(<AddTopic {...props} />);

  return {
    wrapper,
    props
  };
};

describe('AddTopic Component', () => {
  it('render correctly', () => {
    const { wrapper } = setup();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should properly fire `addTopic` callback when submit a new topic', () => {
    const { wrapper, props } = setup();
    const form = wrapper.find('form');
    const textArea = wrapper.find('textarea');
    textArea.node.value = 'This is a test';
    textArea.simulate('change', textArea);
    form.simulate('submit', {
      preventDefault: () => {}
    });
    expect(props.addTopic).toHaveBeenCalled();
  });

  it('should show an error that indicate the topic exceed 255 characters', () => {
    const { wrapper } = setup();
    const errorMsgLabelWrapper = wrapper.find('.error');
    const form = wrapper.find('form');
    const textArea = wrapper.find('textarea');
    textArea.node.value = 'a'.repeat(256);
    textArea.simulate('change', textArea);
    form.simulate('submit', {
      preventDefault: () => {}
    });
    expect(errorMsgLabelWrapper.hasClass('invisible')).toBe(false);

});
});
