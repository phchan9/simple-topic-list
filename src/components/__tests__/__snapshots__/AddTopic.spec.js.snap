// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTopic Component render correctly 1`] = `
<AddTopic
  addTopic={[Function]}
>
  <div>
    <form
      className="form-horizontal"
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="topic"
        >
          Topic:
        </label>
        <textarea
          className="form-control"
          id="topic"
          onChange={[Function]}
          placeholder="A topic doesn't exceed 255 characters."
          rows="5"
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <button
          className="btn btn-default"
          type="submit"
        >
          Submit
        </button>
        <label
          className="error invisible"
          htmlFor="submit"
        >
           * A topic does not exceed 255 characters
        </label>
      </div>
    </form>
  </div>
</AddTopic>
`;
